### DISCUSSION TITLE
When the set operation is executed in the 「then callback」, the 「set operation」 takes nearly a hundred times longer.

### DISCUSSION THREAD

**Discussion by LZY622** [pmndrs/zustand#2114] [Bug report]
## When the set operation is executed in the 「then callback」, the「set operation」 takes nearly a hundred times longer.
In a new umi framework，Add a button to the page and add a handle to the button to test the time consumption of the 「set operation」.

### Print time-consuming code(Store)
```TypeScript
import {create} from "zustand";
export const useStore = create<any>()((set, get) => ({
  loading: false, 
  setLoading: (loading: boolean) => {
    console.time('time--set')
    set({ loading })
    console.timeEnd('time--set')
  },
}))
```

### UI code
```TSX
return (
  <div>
    <div>{String(loading)}</div>
    <Button type="primary" onClick={testSetHandle}>setLoading</Button>
  </div>
);
```

### Experiment 1
Handle code：
```TypeScript
const testSetHandle = useCallback(async () => {
  setLoading(!loading)
}, [loading])
```
Time consuming printing：
![set0](https://github.com/pmndrs/zustand/assets/42869166/027ce227-9f31-4655-afe4-aa911708bb6d)

### Experiment 2
Handle code：
```TypeScript
const testSetHandle = useCallback(async () => {
  await sleep(3000);
  console.log('sleep 3000')
  setLoading(!loading)
}, [loading])
```
Time consuming printing：
![set1](https://github.com/pmndrs/zustand/assets/42869166/2e697f8c-cff1-41f8-ac92-92e19f14ff12)

### Experiment 3
Handle code：
```TypeScript
const testSetHandle = useCallback(async () => {
  await sleep(3000);
  console.log('sleep 3000')
  await sleep(2000);
  console.log('sleep 2000')
  setLoading(!loading)
}, [loading])
```
Time consuming printing：
![set2](https://github.com/pmndrs/zustand/assets/42869166/8ce28fbb-9c53-4f57-bf2d-21e9c86de92f)

### Experiment 4
Handle code：
```TypeScript
const testSetHandle = useCallback(async () => {
  await sleep(3000);
  console.log('sleep 3000')
  await sleep(2000);
  console.log('sleep 2000')
  await sleep(1000);
  console.log('sleep 1000')
  setLoading(!loading)
}, [loading])
```
Time consuming printing：
![set3](https://github.com/pmndrs/zustand/assets/42869166/0a106ec1-71e9-4cbe-bdac-2f82230b8e47)

### Experiment 5
Handle code：
```TypeScript
const testSetHandle = useCallback(async () => {
  await sleep(3000);
  console.log('sleep 3000')
  await sleep(2000);
  console.log('sleep 2000')
  await sleep(1000);
  console.log('sleep 1000')
  await sleep(2000);
  console.log('sleep 2000')
  setLoading(!loading)
}, [loading])
```
Time consuming printing：
![set4](https://github.com/pmndrs/zustand/assets/42869166/8bbc1c15-f82c-456e-9138-e689911e3b9f)

### Conclusions and conjectures
If there is an await statement before the set operation, the set operation will take longer. And the more awaits, the more time it takes. From experiments, we can find that there seems to be a threshold for the number of awaits. In the example, if there are more than 2 awaits, the time consumption will jump directly to 30ms.In our actual project, due to factors such as store size, a simple set operation will take 200+ms. The essence of await is the syntactic sugar of 「.then callback」, so it is reasonable to speculate that it is related to the number of nesting of 「.then」.Please suggest some solutions, thank you very much

## Link to reproduction
The code is relatively simple to reproduce，code repository is no longer provided

## Check List

Please do not ask questions in issues.

- [X] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions/2019) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [ ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [X] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
Hi, thanks for reporting.
Would it change if it's a normal function without Zustand?
Also, what happens if you remove showing `loading` and use the function directly `const { setLoading } = useStore.getState();`?

**Reply by LZY622** [reply]
hello，I removed showing `loading`, and use static method ` const { setLoading } = useStore.getState();` ,but there hasn't been any change. And compared it with the native setState
![set-two1](https://github.com/pmndrs/zustand/assets/42869166/afb80760-6d2e-4856-8749-35397f4820fe)
![set-two2](https://github.com/pmndrs/zustand/assets/42869166/36721d4d-e259-41a0-9bce-be51a7e27ad9)
![set-two3](https://github.com/pmndrs/zustand/assets/42869166/8097aff4-271a-4c38-84bc-c7f642840488)

**Reply by dai-shi** [reply]
Thanks for investigating it.
That's interesting.
I hope someone can dig into it.
Let me add the "help wanted" label.

**Reply by dai-shi** [reply]
> I removed showing `loading`

Ah, wait, can you remove the line `const { loading } = useStore(e => e);` entirely?

> And compared it with the native setState

So, what would be your insight?
